src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportFileManager.cs

```diff
@@ -0,0 +1,134 @@
+    /// has reached a certain size) and keeping track of current file for each resource type.
+    /// Expected use case: each ExportJobTask will instantiate a copy; single-threaded access.
+    /// </summary>
+    internal class ExportFileManager
+    {
+        private readonly ExportJobRecord _exportJobRecord;
+        private readonly IExportDestinationClient _exportDestinationClient;
+        private readonly IDictionary<string, ExportFileInfo> _resourceTypeToFileInfoMapping;
+        private readonly uint _maxFileSizeInBytes;
+        private bool _isInitialized = false;
+
+        public ExportFileManager(ExportJobRecord exportJobRecord, IExportDestinationClient exportDestinationClient)
+        {
+            EnsureArg.IsNotNull(exportJobRecord, nameof(exportJobRecord));
+            EnsureArg.IsNotNull(exportDestinationClient, nameof(exportDestinationClient));
+            EnsureArg.IsGt(exportJobRecord.MaxFileSizeInMB, 0, nameof(exportJobRecord.MaxFileSizeInMB));
+
+            _exportJobRecord = exportJobRecord;
+            _exportDestinationClient = exportDestinationClient;
+            _maxFileSizeInBytes = _exportJobRecord.MaxFileSizeInMB * 1024 * 1024;
+            _resourceTypeToFileInfoMapping = new Dictionary<string, ExportFileInfo>();
+        }
+
+        private async Task Initialize(CancellationToken cancellationToken)
+        {
+            // Each resource type can have multiple files. We need to keep track of the latest file.
+            foreach (KeyValuePair<string, List<ExportFileInfo>> output in _exportJobRecord.Output)
+            {
+                ExportFileInfo latestFile = output.Value[output.Value.Count - 1];
+                _resourceTypeToFileInfoMapping.Add(output.Key, latestFile);
+
+                // If there are entries in ExportJobRecord Output before FileManager gets initialized,
+                // it means we are "restarting" an export job. We have to make sure that each file
+                // has been opened on the ExportDestinationClient.
+                await _exportDestinationClient.OpenFileAsync(latestFile.FileUri, cancellationToken);
+            }
+
+            _isInitialized = true;
+        }
+
+        public async Task WriteToFile(string resourceType, string partId, byte[] data, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNullOrWhiteSpace(resourceType, nameof(resourceType));
+
+            if (!_isInitialized)
+            {
+                await Initialize(cancellationToken);
+            }
+
+            ExportFileInfo fileInfo = await GetFile(resourceType, cancellationToken);
+            await _exportDestinationClient.WriteFilePartAsync(fileInfo.FileUri, partId, data, cancellationToken);
+
+            fileInfo.IncrementCount(data.Length);
+
+            // We need to create a new file if the current file has exceeded a certain limit.
+            if (fileInfo.CommittedBytes >= _maxFileSizeInBytes)
```

Comment:
This makes me concerned about calling the variable maxFileSizeInBytes. Most files will be slightly larger than this value before a new one is made, so if a user sets the value to the max size a storage account file can be we could see the same error we currently get. Would it be possible to check if the next data write would put the file over the limit, and make a new file before we go above this point?

Reply:
This is intended to be more of an "approximate" max file size. My plan was to call this out when we add documentation for this configuration. 

One issue that could happen if we check for file size before adding is that we have the potential to end up in an infinite loop without exporting any data. This could happen if the user sets a very low limit value by mistake. Or if the data that has to be added is greater than the limit set by the user (say 1MB limit and the resource data that needs to be uploaded is 2MB). To handle this scenario we have to add more logic to "force" add files when we encounter such situations. We finally end up with a file size that is greater than the specified limit. We will have to call this out in the documentation either way.

I'll update the code to check before adding.

Reply:
That is a good point. I'm ok with the current approach and handling this in documentation.